WebDriverAgentLib/Utilities/FBImageIOScaler.m

```diff
@@ -55,51 +65,105 @@ - (void)submitImage:(NSData *)image scalingFactor:(CGFloat)scalingFactor compres
-                                                       (const NSString *)kCGImageSourceCreateThumbnailFromImageIfAbsent: @(YES),
-                                                       (const NSString *)kCGImageSourceThumbnailMaxPixelSize: @(scaledMaxPixelSize)
-                                                       };
-
+    (const NSString *)kCGImageSourceCreateThumbnailWithTransform: @(YES),
+    (const NSString *)kCGImageSourceCreateThumbnailFromImageIfAbsent: @(YES),
+    (const NSString *)kCGImageSourceThumbnailMaxPixelSize: @(scaledMaxPixelSize)
+  };
   CGImageRef scaled = CGImageSourceCreateThumbnailAtIndex(imageData, 0, params);
-  if (scaled == nil) {
-    [FBLogger log:@"Failed to scale the image"];
-    CFRelease(imageData);
+  CFRelease(imageData);
+  if (nil == scaled) {
+    [[[FBErrorBuilder builder]
+      withDescriptionFormat:@"Failed to scale the image"]
+     buildError:error];
     return nil;
   }
-  NSData *jpegData = [self jpegDataWithImage:scaled
-                           compressionQuality:compressionQuality];
+  NSData *resData = [self jpegDataWithImage:scaled
+                         compressionQuality:compressionQuality];
+  if (nil == resData) {
+    [[[FBErrorBuilder builder]
+      withDescriptionFormat:@"Failed to compress the image to JPEG format"]
+     buildError:error];
+  }
   CGImageRelease(scaled);
-  CFRelease(imageData);
-  return jpegData;
+  return resData;
 }
 
-- (nullable NSData *)jpegDataWithImage:(CGImageRef)imageRef compressionQuality:(CGFloat)compressionQuality
+- (nullable NSData *)scaledImageWithImage:(NSData *)image
+                                      uti:(NSString *)uti
+                                     rect:(CGRect)rect
+                            scalingFactor:(CGFloat)scalingFactor
+                       compressionQuality:(CGFloat)compressionQuality
+                                    error:(NSError **)error
+{
+  UIImage *uiImage = [UIImage imageWithData:image];
+  CGSize size = uiImage.size;
+  CGSize scaledSize = CGSizeMake(size.width * scalingFactor, size.height * scalingFactor);
+  UIGraphicsBeginImageContext(scaledSize);
+  UIImageOrientation orientation = uiImage.imageOrientation;
+#if !TARGET_OS_TV
+  if (FBConfiguration.screenshotOrientation == UIInterfaceOrientationPortrait) {
+    orientation = UIImageOrientationUp;
+  } else if (FBConfiguration.screenshotOrientation == UIInterfaceOrientationPortraitUpsideDown) {
+    orientation = UIImageOrientationDown;
+  } else if (FBConfiguration.screenshotOrientation == UIInterfaceOrientationLandscapeLeft) {
+    orientation = UIImageOrientationRight;
+  } else if (FBConfiguration.screenshotOrientation == UIInterfaceOrientationLandscapeRight) {
+    orientation = UIImageOrientationLeft;
+  }
+#endif
+  uiImage = [UIImage imageWithCGImage:(CGImageRef)uiImage.CGImage
+                                scale:uiImage.scale
```

Comment:
Should here be `1.0`?
I ran the failing view test with WDA 3.4.0, then it was passed. Main point was different image width/height.
https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=14589&view=results

diff: https://github.com/appium/appium-xcuitest-driver/compare/get-viewport-screenshot-check?expand=1

(I haven't tested with this 1.0)

Reply:
I did check the actual screenshot file resolution and it was ok. Perhaps, there is a bug in the images parsing/comparison? I assume the image library there must properly handle scaling if it is present in the image metadata.

Reply:
mm, thanks for the checking.
noted https://github.com/appium/WebDriverAgent/issues/487